### PR TITLE
Add tests

### PR DESCRIPTION
**Pull request by Yeti-or**
resolves: #19 

@zxqfox take a look on tests structure please

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+sudo: false
+
+language: node_js
+
+matrix:
+  include:
+    - node_js: "8"
+    - node_js: "9"
+      env: COVERALLS=1
```
Comment (zxqfox) [inline .travis.yml]: Почему coveralls на нестабильной?

File: test/helpers/mock-fs.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    'path/to/some.png': new Buffer([8, 6, 7, 5, 3, 0, 9]),
+    'some/other/path': {/ another empty directory /}
+  });
+  ```
+*/
+
+const createFile = (fs, path, content) => fs.writeFileSync(path, content);
+const createDir = (fs, path) => fs.mkdirpSync(path);
+
+function popuplate(fs, paths, currentDir) {
```
Comment (zxqfox) [inline test/helpers/mock-fs.js]: popup late?

File: test/helpers/compiler.js [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+module.exports = function(fixture, { config, mock }) {
+    const testDir = process.cwd();
+    const output = {
+        path : path.resolve('output'),
+        filename : '[name].bundle.js'
+    };
+
+    // webpack Config
+    config = {
```
Comment (zxqfox) [inline test/helpers/compiler.js]: а зачем ты переопределяешь его?

File: test/options/levels.test.js [hunk 4]
```diff
@@ -0,0 +1,37 @@
+const webpack = require('../helpers/compiler');
+
+describe('Options', () => {
+    describe('levels', () => {
+        test('common/desktop', async () => {
+            const mock = {
+                'index.js' : `require('b:button')`,
+                'common.blocks' : {
+                    'button' : {
+                        'button.js' : `({ block: 'button', content: 'common' })`
```
Comment (zxqfox) [inline test/options/levels.test.js]: почему не `common.blocks/button/button.js' : '?'`

File: test/options/techs.test.js [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                }
+            };
+
+            const { stats } = await webpack('index.ts', { config, mock });
+            const { source } = stats.toJson().modules[1];
+
+            expect(source).toMatchSnapshot();
+        });
+
+        // TODO: issue/
```
Comment (zxqfox) [inline test/options/techs.test.js]: issue

File: test/options/techs.test.js [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+            };
+
+            const { stats } = await webpack('index.ts', { config, mock });
+            const { source } = stats.toJson().modules[2];
+
+            console.log(source);
+            // expect(source).toMatchSnapshot();
+        });
+
+        // TODO: issue/
```
Comment (zxqfox) [inline test/options/techs.test.js]: issue?

File: test/options/techs.test.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        // We need to choose one if we have both extensions on one level
+        test.skip('ts & js on same levels', async () => {
+            const mock = {
+                'index.ts' : `require('b:button')`,
+                'common.blocks' : {
+                    'button' : {
+                        'button.ts' : `({ block: 'button' })`,
+                        'button.js' : `({ block: 'button', content: 'I love js' })`
```
Comment (zxqfox) [inline test/options/techs.test.js]: This should be an error, actually, isn't it?

File: test/options/techs.test.js [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+                            js : ['ts']
+                        }
+                    }
+                }
+            };
+
+            const { stats } = await webpack('index.ts', { config, mock });
+            const { source } = stats.toJson().modules[1];
+
+            expect(source).toMatchSnapshot();
```
Comment (zxqfox) [inline test/options/techs.test.js]: Do you really like how this happens?
When you reading tests you don't see what's going on. You have to go to the some file named as concatenation of `e('Options', () => { describe('techs && techMap', () => { test('only ts', ` and then go back to compare with code. It's pretty annoying and uncomfortable.

File: test/order/order.test.js [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        const cssFile = assets['_index.css'];
+        expect(getOrder(cssFile)).toEqual([0, 1]);
+        expect(cssFile).toMatchSnapshot();
+    });
+
+    test('order of modifiers', async () => {
+        const mock = {
+            'index.js' : `require('b:button m:theme=normal|action m:size=m')`,
+            'common.blocks' : {
+                'button' : {
```
Comment (zxqfox) [inline test/order/order.test.js]: `'common.blocks/button' : {`

File: test/order/order.test.js [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+                    'button.css' : `.button { order: 0 }\n`,
+                    '_theme' : {
+                        'button_theme_normal.css' : `.button_theme_normal { order: 1 }\n`,
```
Comment (zxqfox) [inline test/order/order.test.js]: `'_theme/button_theme_normal.css' : `?

File: test/order/order.test.js [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+                        'button_size_m.css' : `.button_size_m, { order: 3 }\n`
+                    }
+                }
+            }
+        };
+        const { assets } = await webpack('index.js', { config, mock });
+
+        const cssFile = assets['_index.css'];
+        expect(getOrder(cssFile)).toEqual([0, 1, 2, 3]);
+        expect(cssFile).toMatchSnapshot();
```
Comment (zxqfox) [inline test/order/order.test.js]: why 2 asserts about the same?

File: test/options/techs.test.js (hunk 7 above)
Comment (Yeti-or) [reply test/options/techs.test.js]: why?  what do you mean ?

File: test/options/techs.test.js (hunk 8 above)
Comment (Yeti-or) [reply test/options/techs.test.js]: actually, I agree with you. But it's you know jest/hype snapshot testing. And I took an example from guys who write webpack its self.
Do you think it's better to assert strings ?

File: .travis.yml (hunk 1 above)
Comment (Yeti-or) [reply .travis.yml]: whats the difference?

Comment (zxqfox) [reply .travis.yml]: no matter what. just no need to pull troubles out of nothing ;-)

File: index.js [hunk 12]
```diff
@@ -76,21 +74,22 @@ module.exports = function(source) {
...
-                        cell : bemCell,
-                        exist,
-                        // prepare path for require cause relative returns us string that we couldn't require
-                        path : unifyPath(requiredPath(path.relative(path.dirname(this.resourcePath), entityPath)))
-                    };
-                });
+            return new Promise(resolve => this.fs.stat(entityPath, err => {
+                return resolve(!err);
+            }))
+            .then(exist => {
```
Comment (zxqfox) [inline index.js]: Each `then` produces new Promise object.

Comment (Yeti-or) [reply index.js]: It was minimum diff change
I fixed it

**Pull request merged**